### ISSUE TITLE
pg_upgrade error

### ISSUE THREAD

**Issue by ransom85** [orafce/orafce#124]
I'm trying to upgrade from version 12 to version 13.

I installed the server including the extension:
yum install postgresql13-server
yum install postgresql13-contrib
yum install orafce13
yum install plpgsql_check_13

First I tried to verify the possibility of an upgrade, everything turned out fine:
/usr/pgsql-13/bin/pg_upgrade --check --old-datadir /u02/12/ --new-datadir /u02/13/ --old-bindir /usr/pgsql-12/bin/ --new-bindir /usr/pgsql-13/bin/

Provádím Kontrolu Konzistence na Starém Live Serveru
----------------------------------------------------
Checking cluster versions                                   ok
Checking database user is the install user                  ok
Checking database connection settings                       ok
Checking for prepared transactions                          ok
Checking for reg* data types in user tables                 ok
Checking for contrib/isn with bigint-passing mismatch       ok
Checking for presence of required libraries                 ok
Checking database user is the install user                  ok
Checking for prepared transactions                          ok
Checking for new cluster tablespace directories             ok

*Clustery jsou kompatibilní*

But the following error occurred during the upgrade:

Performing Upgrade
------------------
Analyzing all rows in the new cluster                       ok
Freezing all rows in the new cluster                        ok
Deleting files from new pg_xact                             ok
Copying old pg_xact to new server                           ok
Setting next transaction ID and epoch for new cluster       ok
Deleting files from new pg_multixact/offsets                ok
Copying old pg_multixact/offsets to new server              ok
Deleting files from new pg_multixact/members                ok
Copying old pg_multixact/members to new server              ok
Setting next multixact ID and offset for new cluster        ok
Resetting WAL archives                                      ok
Setting frozenxid and minmxid counters in new cluster       ok
Restoring global objects in the new cluster                 ok
Restoring database schemas in the new cluster
  template1
*failure*

Consult the last few lines of "pg_upgrade_dump_1.log" for
the probable cause of the failure.
Failure, exiting

It is listed in the log:
pg_restore: creating FUNCTION "public.nvarchar2("public"."nvarchar2", integer, boolean)"
pg_restore: while PROCESSING TOC:
pg_restore: from TOC entry 611; 1255 17575 FUNCTION nvarchar2("public"."nvarchar2", integer, boolean) postgres
pg_restore: error: could not execute query: ERROR:  function public.nvarchar2_transform(internal) does not exist
Command was: CREATE FUNCTION "public"."nvarchar2"("public"."nvarchar2", integer, boolean) RETURNS "public"."nvarchar2"
    LANGUAGE "c" IMMUTABLE STRICT SUPPORT "public"."nvarchar2_transform"
    AS '$libdir/orafce', 'nvarchar2';

-- For binary upgrade, handle extension membership the hard way
ALTER EXTENSION "orafce" ADD FUNCTION "public"."nvarchar2"("public"."nvarchar2", integer, boolean);

Thanks for the help

**Comment by okbob** [comment]
po 21. 12. 2020 v 8:27 odesílatel ransom85 <notifications@github.com>
napsal:

> I'm trying to upgrade from version 12 to version 13.
>
> I installed the server including the extension:
> yum install postgresql13-server
> yum install postgresql13-contrib
> yum install orafce13
> yum install plpgsql_check_13
>
> First I tried to verify the possibility of an upgrade, everything turned
> out fine:
> /usr/pgsql-13/bin/pg_upgrade --check --old-datadir /u02/12/ --new-datadir
> /u02/13/ --old-bindir /usr/pgsql-12/bin/ --new-bindir /usr/pgsql-13/bin/
> Provádím Kontrolu Konzistence na Starém Live Serveru
>
> Checking cluster versions ok
> Checking database user is the install user ok
> Checking database connection settings ok
> Checking for prepared transactions ok
> Checking for reg* data types in user tables ok
> Checking for contrib/isn with bigint-passing mismatch ok
> Checking for presence of required libraries ok
> Checking database user is the install user ok
> Checking for prepared transactions ok
> Checking for new cluster tablespace directories ok
>
> *Clustery jsou kompatibilní*
>
> But the following error occurred during the upgrade:
> Performing Upgrade
>
> Analyzing all rows in the new cluster ok
> Freezing all rows in the new cluster ok
> Deleting files from new pg_xact ok
> Copying old pg_xact to new server ok
> Setting next transaction ID and epoch for new cluster ok
> Deleting files from new pg_multixact/offsets ok
> Copying old pg_multixact/offsets to new server ok
> Deleting files from new pg_multixact/members ok
> Copying old pg_multixact/members to new server ok
> Setting next multixact ID and offset for new cluster ok
> Resetting WAL archives ok
> Setting frozenxid and minmxid counters in new cluster ok
> Restoring global objects in the new cluster ok
> Restoring database schemas in the new cluster
> template1
> *failure*
>
> Consult the last few lines of "pg_upgrade_dump_1.log" for
> the probable cause of the failure.
> Failure, exiting
>
> It is listed in the log:
> pg_restore: creating FUNCTION "public.nvarchar2("public"."nvarchar2",
> integer, boolean)"
> pg_restore: while PROCESSING TOC:
> pg_restore: from TOC entry 611; 1255 17575 FUNCTION
> nvarchar2("public"."nvarchar2", integer, boolean) postgres
> pg_restore: error: could not execute query: ERROR: function
> public.nvarchar2_transform(internal) does not exist
> Command was: CREATE FUNCTION "public"."nvarchar2"("public"."nvarchar2",
> integer, boolean) RETURNS "public"."nvarchar2"
> LANGUAGE "c" IMMUTABLE STRICT SUPPORT "public"."nvarchar2_transform"
> AS '$libdir/orafce', 'nvarchar2';
>
> -- For binary upgrade, handle extension membership the hard way
> ALTER EXTENSION "orafce" ADD FUNCTION
> "public"."nvarchar2"("public"."nvarchar2", integer, boolean);
>
Unfortunately, It looks like pg_dump bug.

I'll try report this issue

Regards

Pavel

**Comment by okbob** [comment]
Unfortunately - it was old Orafce bug. It will be fixed on Orafce 3.14.

For older releases there is workaround. You should to run following INSERTs (as superuser)
<pre>
  INSERT INTO pg_depend (classid, objid, objsubid,
                           refclassid, refobjid, refobjsubid, deptype)
       VALUES('pg_proc'::regclass::oid, 'varchar2'::regproc::oid, 0,
              'pg_proc'::regclass::oid, 'varchar2_transform'::regproc::oid, 0, 'n');

    INSERT INTO pg_depend (classid, objid, objsubid,
                           refclassid, refobjid, refobjsubid, deptype)
       VALUES('pg_proc'::regclass::oid, 'nvarchar2'::regproc::oid, 0,
              'pg_proc'::regclass::oid, 'nvarchar2_transform'::regproc::oid, 0, 'n');
</pre>

It should to fix pg_upgrade.

**Comment by ukolovda** [comment]
Thank you!

I solved this problem after upgrading Fedora 33 to 34:

1. install PostgreSQL 12 ( https://www.postgresql.org/download/linux/redhat/ )
2. Start the server in console (enter your path to data): 
```
# su -l postgres
$ PGDATA=/var/lib/pgsql/data /usr/pgsql-12/bin/postgres
```
3. In second console drop the extension (use correct database name in `psql` command, repeat for every failure database):
```
# su -l postgres
$ psql <database>
postgres=# drop extension orafce;
DROP EXTENSION
postgres=# \q
```
You also can inserts rows as described in workaround above.

4. Stop the 12 server in first console.
5. Run `postgresql-setup --upgrade`. It complete ok.
6. Start service as usual.
7. Uninstall PostgreSQL 12.
